std/traits.d

```diff
@@ -6839,3 +6839,344 @@ unittest
+    }
+    else
+        enum isEnum = false;
+}
+
+///
+unittest
+{
+    enum E1 { a, b }
+    enum E2 : string { a = "A", b = "B" }
+    static assert(isEnum!E1);
+    static assert(isEnum!E2);
+}
+
+version(unittest)
+{
+    // for module variables
+    int mx;
+    immutable int ix = 0;
+    immutable int ix2;
+    enum int ex = 0;
+    static assert( isVariable!mx);
+    static assert( isVariable!ix);
+    static assert( isVariable!ix2);
+    static assert(!isVariable!ex);
+    static assert(!isEnum!mx);
+    static assert(!isEnum!ix);
+    static assert(!isEnum!ix2);
+    static assert( isEnum!ex);
+}
+unittest
+{
+    // for local variables
+    int my;
+    immutable int iy = 0;
+    enum int ey = 0;
+    static assert( isVariable!my);
+    static assert( isVariable!iy);
+    static assert(!isVariable!ey);
+    static assert(!isEnum!my);
+    static assert(!isEnum!iy);
+    static assert( isEnum!ey);
+
+    // for member variables
+    struct S {
+        int mz;
+        immutable int iz = 0;
+        immutable int iz2;  this(int) { iz2 = 0; }
+        enum int ez = 0;
+
+        static int sval;
+
+        template Test() {}
+    }
+    static assert( isVariable!(S.mz));
+    static assert( isVariable!(S.iz));
+    static assert( isVariable!(S.iz2));
+    static assert(!isVariable!(S.ez));
+    static assert( isVariable!(S.sval));
+    static assert(!isEnum!(S.mz));
+    static assert(!isEnum!(S.iz));
+    static assert(!isEnum!(S.iz2));
+    static assert( isEnum!(S.ez));
+    static assert(!isEnum!(S.sval));
+
+    void function() fp;
+    void delegate() dg;
+    static assert( isVariable!(fp));
+    static assert( isVariable!(dg));
+    static assert(!isEnum!(fp));
+    static assert(!isEnum!(dg));
+
+    // for function and property function
+    void func(){}
+    @property int prop(){ return 0; }
+    static assert(!isVariable!func);
+    static assert(!isVariable!prop);
+    static assert(!isEnum!func);
+    static assert(!isEnum!prop);
+
+    // for aggregates
+    class C {}
+    interface I {}
+    union U {}
+    static assert(!isVariable!S);
+    static assert(!isVariable!C);
+    static assert(!isVariable!I);
+    static assert(!isVariable!U);
+    static assert(!isEnum!S);
+    static assert(!isEnum!C);
+    static assert(!isEnum!I);
+    static assert(!isEnum!U);
+
+    // for other symbols
+    static assert(!isVariable!(S.Test));        // template
+    static assert(!isVariable!(S.Test!()));     // template instance
+    static assert(!isEnum!(S.Test));            // template
+    static assert(!isEnum!(S.Test!()));         // template instance
+
+    enum E { x = 10 }
+    static assert(!isVariable!(E.x));
+    static assert( isEnum!(E.x));
+}
+
+/**
+ * Detect whether $(D X) is a final function or class.
+ *
+ * Returns:
+ *     `bool`
+ */
+template isFinal(X...) if (X.length == 1)
+{
+    static if (is(X[0] == class))
+        enum isFinal = __traits(isFinalClass, X[0]);
+    else static if (isSomeFunction!X)
```

Comment:
Should be `isFunction`, I don't think function pointers or delegates have `final`